Objects/listobject.c

```diff
@@ -31,6 +31,49 @@ get_list_freelist(void)
 }
 #endif
 
+#ifdef Py_GIL_DISABLED
+static _PyListArray *
+list_allocate_array(size_t capacity)
+{
+    if (capacity > PY_SSIZE_T_MAX/sizeof(PyObject*) - 1) {
+        return NULL;
+    }
+    _PyListArray *array = PyMem_Malloc(sizeof(_PyListArray) + (capacity - 1) * sizeof(PyObject *));
+    if (array == NULL) {
+        return NULL;
+    }
+    array->allocated = capacity;
+    return array;
+}
+
+static Py_ssize_t
+list_capacity(PyObject **items)
+{
+    char *mem = (char *)items;
+    mem -= offsetof(_PyListArray, ob_item);
+    _PyListArray *array = (_PyListArray *)mem;
+    return _Py_atomic_load_ssize_relaxed(&array->allocated);
```

Comment:
Hmm by the way, do we need to use `_Py_atomic_load_ssize_relaxed ` here?
IIUC, we always have a copy of array and it's immutable.

Reply:
Yeah, we shouldn't need the atomic load.